quinn-proto/src/stream.rs

```diff
@@ -400,97 +399,98 @@ pub enum RecvState {
+        let mut read = 0;
+        loop {
+            if self.consume(buf, &mut read) {
+                self.data.pop();
+            } else {
+                break;
+            }
+            if read == buf.len() {
+                break;
             }
-            return (i * 8 + x.leading_zeros() as usize) - self.written_offset as usize;
         }
-        self.written.len()
+        read
     }
 
-    pub fn read(&mut self, buf: &mut [u8]) -> usize {
-        let (a, b) = self.data.as_slices();
-        let available = self.prefix_len();
-        let a_len = a.len().min(available);
-        let (a, b) = (&a[0..a_len], &b[0..(available - a_len)]);
-        let a_n = a.len().min(buf.len());
-        buf[0..a_n].copy_from_slice(&a[0..a_n]);
-        let b_n = b.len().min(buf.len().saturating_sub(a.len()));
-        buf[a_n..(a_n + b_n)].copy_from_slice(&b[0..b_n]);
-        let n = a_n + b_n;
-
-        self.offset += n as u64;
-        self.data.drain(0..n);
-        let q = n / 8;
-        let r = n % 8;
-        let carry = (self.written_offset as usize + r) / 8;
-        self.written.drain(0..(q + carry));
-        self.written_offset = (self.written_offset as usize + r - carry * 8) as u8;
-
-        n
+    // Read as much from the first chunk in the heap as fits in the buffer.
+    // Takes the buffer to read into and the amount of bytes that has already
+    // been read into it. Returns whether the first chunk has been fully consumed.
+    fn consume(&mut self, buf: &mut [u8], read: &mut usize) -> bool {
+        let mut chunk = if let Some(chunk) = self.data.peek_mut() {
+            chunk
+        } else {
+            return false;
+        };
+
+        // If this chunk is either after the current offset or fully before it,
+        // return directly, indicating whether the chunk can be discarded.
+        if chunk.offset > self.offset {
+            return false;
+        } else if (chunk.offset + chunk.bytes.len() as u64) < self.offset {
+            return true;
+        }
+
+        // Determine `start` and `len` of slice to read from chunk
+        let start = if chunk.offset < self.offset {
+            (self.offset - chunk.offset) as usize
+        } else {
+            0
+        };
+        let left = buf.len() - *read;
+        let len = left.min(chunk.bytes.len() - start) as usize;
+
+        // Actually write into the buffer and update the related state
+        (&mut buf[*read..*read + len]).copy_from_slice(&chunk.bytes[start..start + len]);
+        *read += len;
+        self.offset += len as u64;
+
+        return if start + len == chunk.bytes.len() {
+            // This chunk has been fully consumed and can be discarded
+            true
+        } else {
+            // Mutate the chunk; per the documentation, this will update the
+            // heap's ordering accordingly if necessary. Don't pop the chunk.
+            chunk.offset = chunk.offset + start as u64 + len as u64;
```

Comment:
Where does the heap get reordered?

Reply:
Presumably on dropping the `PeekMut`. It's documented for `peek_mut()`, I've tried to clarify the comment.